calculator/main.py

```diff
@@ -26,7 +33,29 @@ def perform_operation(calc, operation, params):
     :param params: Tuple containing the list of nums to operate with.
                    ie: (1, 2, 3, 4.5, -2)
     """
-    pass
+    # Check to see if the operation exists
+    # Enabling this requires enabling the test suite:
+    # test_perform_nonexistant_operation
+    '''
+    if operation not in get_operations(calc):
+        raise InvalidOperation('Given operation is invalid.')
+    '''
+    
+    # Check the params
+    for num in params:
+        if not isinstance(num,(int, float, complex)):
+            raise InvalidParams('Given params are invalid.')
+
+    # compute the result
+    result =  calc['operations'][operation](*params)
+    
+    #Update the history with the operation.
+    #logentry format: ('2016-05-20 12:00:00', 'add', (1, 2), 3)
+    timestamp = '{:%Y-%m-%d %H:%M:%S}'.format(datetime.now())
```

Comment:
The way you did it works, but there's also a strftime format functionality you might want to look into for later.